zfs-2.2.3-staging does not compile on FreeBSD

### System information
<!--  add version after "|" character -->
Type | Version/Name
 --- | ---
Distribution Name	| FreeBSD
Distribution Version	| 14.0-STABLE
Kernel Version	| 38fe4bc08264652cd33964ba6042bd4c0faf6b76
Architecture	| amd64
OpenZFS Version	| 2.2.3-staging

### Describe the problem you're observing
After backport of 6930ecbb75e17f6d6cc7686211d4f3b34852795b (#15675) does zfs-2.2.3-staging not compile on FreeBSD 14.

### Describe how to reproduce the problem
Try to compile zfs-2.2.3-staging on FreeBSD 14

### Include any warning/errors/backtraces from the system logs
Error output:
```
cc  -O2 -pipe  -std=gnu99 -Wno-format-zero-length -nobuiltininc -idirafter /usr/lib/clang/16/include -fstack-protector-strong -Wall -Wstrict-prototypes -Wmissing-prototypes -Wpointer-arith -Wcast-qual -Wundef -Wno-pointer-sign -D__printf__=__freebsd_kprintf__ -Wmissing-include-dirs -fdiagnostics-show-option -Wno-unknown-pragmas -Wno-error=tautological-compare -Wno-error=empty-body -Wno-error=parentheses-equality -Wno-error=unused-function -Wno-error=pointer-sign -Wno-error=shift-negative-value -Wno-address-of-packed-member -Wno-format-zero-length   -Qunused-arguments -Wno-cast-qual   -include /git/zfs/include/os/freebsd/spl/sys/ccompile.h -D__KERNEL__ -DFREEBSD_NAMECACHE -DBUILDING_ZFS -D__BSD_VISIBLE=1 -DHAVE_UIO_ZEROCOPY -DWITHOUT_NETDUMP -D__KERNEL -D_SYS_CONDVAR_H_ -D_SYS_VMEM_H_ -DKDTRACE_HOOKS -DCOMPAT_FREEBSD11 -D__x86_64 -DHAVE_SSE2 -DHAVE_SSSE3 -DHAVE_SSE4_1 -DHAVE_SSE4_2 -DHAVE_AVX -DHAVE_AVX2 -DHAVE_AVX512F -DHAVE_AVX512VL -DHAVE_AVX512BW -DNDEBUG -DBITS_PER_LONG=64 -fno-strict-aliasing -Werror -D_KERNEL -DKLD_MODULE -nostdinc  -I/git/zfs/include -I/git/zfs/module/icp/include -I/git/zfs/include/os/freebsd -I/git/zfs/include/os/freebsd/spl -I/git/zfs/include/os/freebsd/zfs -I/git/zfs/module/zstd/include -I/git/zfs/module -include /git/zfs/module/opt_global.h -I. -I/usr/src/sys -I/usr/src/sys/contrib/ck/include -fno-common -g -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -fdebug-prefix-map=./machine=/usr/src/sys/amd64/include -fdebug-prefix-map=./x86=/usr/src/sys/x86/include -fdebug-prefix-map=./i386=/usr/src/sys/i386/include     -MD  -MF.depend.spa.o -MTspa.o -mcmodel=kernel -mno-red-zone -mno-mmx -mno-sse -msoft-float  -fno-asynchronous-unwind-tables -ffreestanding -fwrapv -fstack-protector -Wall -Wstrict-prototypes -Wmissing-prototypes -Wpointer-arith -Wcast-qual -Wundef -Wno-pointer-sign -D__printf__=__freebsd_kprintf__ -Wmissing-include-dirs -fdiagnostics-show-option -Wno-unknown-pragmas -Wno-error=tautological-compare -Wno-error=empty-body -Wno-error=parentheses-equality -Wno-error=unused-function -Wno-error=pointer-sign -Wno-error=shift-negative-value -Wno-address-of-packed-member -Wno-format-zero-length   -Qunused-arguments -mno-aes -mno-avx -Wno-cast-qual -std=gnu99 -c /git/zfs/module/zfs/spa.c -o spa.o
/git/zfs/module/zfs/spa.c:10501:1: error: too few arguments provided to function-like macro invocation
ZFS_MODULE_VIRTUAL_PARAM_CALL(zfs_zio, zio_, taskq_read,
^
/git/zfs/include/os/freebsd/spl/sys/mod_os.h:53:39: note: expanded from macro 'ZFS_MODULE_VIRTUAL_PARAM_CALL'
#define ZFS_MODULE_VIRTUAL_PARAM_CALL ZFS_MODULE_PARAM_CALL
                                      ^
/git/zfs/include/os/freebsd/spl/sys/mod_os.h:50:2: note: expanded from macro 'ZFS_MODULE_PARAM_CALL'
        ZFS_MODULE_PARAM_CALL_IMPL(_vfs_ ## scope_prefix, name, perm, \
        ^
/git/zfs/include/os/freebsd/spl/sys/mod_os.h:46:75: note: expanded from macro 'ZFS_MODULE_PARAM_CALL_IMPL'
    SYSCTL_PROC(parent, OID_AUTO, name, CTLFLAG_MPSAFE | perm | args, desc)
                                                                          ^
/usr/src/sys/sys/sysctl.h:815:9: note: macro 'SYSCTL_PROC' defined here
#define SYSCTL_PROC(parent, nbr, name, access, ptr, arg, handler, fmt, descr) \
        ^
/git/zfs/module/zfs/spa.c:10501:1: error: type specifier missing, defaults to 'int'; ISO C99 and later do not support implicit int [-Werror,-Wimplicit-int]
ZFS_MODULE_VIRTUAL_PARAM_CALL(zfs_zio, zio_, taskq_read,
^
/git/zfs/include/os/freebsd/spl/sys/mod_os.h:53:39: note: expanded from macro 'ZFS_MODULE_VIRTUAL_PARAM_CALL'
#define ZFS_MODULE_VIRTUAL_PARAM_CALL ZFS_MODULE_PARAM_CALL
                                      ^
/git/zfs/include/os/freebsd/spl/sys/mod_os.h:50:2: note: expanded from macro 'ZFS_MODULE_PARAM_CALL'
        ZFS_MODULE_PARAM_CALL_IMPL(_vfs_ ## scope_prefix, name, perm, \
        ^
/git/zfs/include/os/freebsd/spl/sys/mod_os.h:46:5: note: expanded from macro 'ZFS_MODULE_PARAM_CALL_IMPL'
    SYSCTL_PROC(parent, OID_AUTO, name, CTLFLAG_MPSAFE | perm | args, desc)
    ^
/git/zfs/module/zfs/spa.c:10504:1: error: too few arguments provided to function-like macro invocation
ZFS_MODULE_VIRTUAL_PARAM_CALL(zfs_zio, zio_, taskq_write,
^
/git/zfs/include/os/freebsd/spl/sys/mod_os.h:53:39: note: expanded from macro 'ZFS_MODULE_VIRTUAL_PARAM_CALL'
#define ZFS_MODULE_VIRTUAL_PARAM_CALL ZFS_MODULE_PARAM_CALL
                                      ^
/git/zfs/include/os/freebsd/spl/sys/mod_os.h:50:2: note: expanded from macro 'ZFS_MODULE_PARAM_CALL'
        ZFS_MODULE_PARAM_CALL_IMPL(_vfs_ ## scope_prefix, name, perm, \
        ^
/git/zfs/include/os/freebsd/spl/sys/mod_os.h:46:75: note: expanded from macro 'ZFS_MODULE_PARAM_CALL_IMPL'
    SYSCTL_PROC(parent, OID_AUTO, name, CTLFLAG_MPSAFE | perm | args, desc)
                                                                          ^
/usr/src/sys/sys/sysctl.h:815:9: note: macro 'SYSCTL_PROC' defined here
#define SYSCTL_PROC(parent, nbr, name, access, ptr, arg, handler, fmt, descr) \
        ^
/git/zfs/module/zfs/spa.c:10504:1: error: type specifier missing, defaults to 'int'; ISO C99 and later do not support implicit int [-Werror,-Wimplicit-int]
ZFS_MODULE_VIRTUAL_PARAM_CALL(zfs_zio, zio_, taskq_write,
^
/git/zfs/include/os/freebsd/spl/sys/mod_os.h:53:39: note: expanded from macro 'ZFS_MODULE_VIRTUAL_PARAM_CALL'
#define ZFS_MODULE_VIRTUAL_PARAM_CALL ZFS_MODULE_PARAM_CALL
                                      ^
/git/zfs/include/os/freebsd/spl/sys/mod_os.h:50:2: note: expanded from macro 'ZFS_MODULE_PARAM_CALL'
        ZFS_MODULE_PARAM_CALL_IMPL(_vfs_ ## scope_prefix, name, perm, \
        ^
/git/zfs/include/os/freebsd/spl/sys/mod_os.h:46:5: note: expanded from macro 'ZFS_MODULE_PARAM_CALL_IMPL'
    SYSCTL_PROC(parent, OID_AUTO, name, CTLFLAG_MPSAFE | perm | args, desc)
    ^
4 errors generated.
*** Error code 1
```

--- a/include/os/freebsd/spl/sys/mod_os.h
+++ b/include/os/freebsd/spl/sys/mod_os.h
@@ -91,6 +91,12 @@
 #define	param_set_max_auto_ashift_args(var) \
     CTLTYPE_UINT, NULL, 0, param_set_max_auto_ashift, "IU"
 
+#define	spa_taskq_read_param_set_args(var) \
+    CTLTYPE_STRING, NULL, 0, spa_taskq_read_param, "A"
+
+#define	spa_taskq_write_param_set_args(var) \
+    CTLTYPE_STRING, NULL, 0, spa_taskq_write_param, "A"
+
 #define	fletcher_4_param_set_args(var) \
     CTLTYPE_STRING, NULL, 0, fletcher_4_param, "A"
 

--- a/module/zfs/spa.c
+++ b/module/zfs/spa.c
@@ -1398,8 +1398,6 @@ spa_taskq_write_param_get(char *buf, zfs_kernel_param_t *kp)
 	return (spa_taskq_param_get(ZIO_TYPE_WRITE, buf));
 }
 #else
-#include <sys/sbuf.h>
-
 /*
  * On FreeBSD load-time parameters can be set up before malloc() is available,
  * so we have to do all the parsing work on the stack.
@@ -1410,19 +1408,11 @@ static int
 spa_taskq_read_param(ZFS_MODULE_PARAM_ARGS)
 {
 	char buf[SPA_TASKQ_PARAM_MAX];
-	int err = 0;
+	int err;
 
-	if (req->newptr == NULL) {
-		int len = spa_taskq_param_get(ZIO_TYPE_READ, buf);
-		struct sbuf *s = sbuf_new_for_sysctl(NULL, NULL, len+1, req);
-		sbuf_cpy(s, buf);
-		err = sbuf_finish(s);
-		sbuf_delete(s);
-		return (err);
-	}
-
+	(void) spa_taskq_param_get(ZIO_TYPE_READ, buf);
 	err = sysctl_handle_string(oidp, buf, sizeof (buf), req);
-	if (err)
+	if (err || req->newptr == NULL)
 		return (err);
 	return (spa_taskq_param_set(ZIO_TYPE_READ, buf));
 }
@@ -1431,19 +1421,11 @@ static int
 spa_taskq_write_param(ZFS_MODULE_PARAM_ARGS)
 {
 	char buf[SPA_TASKQ_PARAM_MAX];
-	int err = 0;
+	int err;
 
-	if (req->newptr == NULL) {
-		int len = spa_taskq_param_get(ZIO_TYPE_WRITE, buf);
-		struct sbuf *s = sbuf_new_for_sysctl(NULL, NULL, len+1, req);
-		sbuf_cpy(s, buf);
-		err = sbuf_finish(s);
-		sbuf_delete(s);
-		return (err);
-	}
-
+	(void) spa_taskq_param_get(ZIO_TYPE_WRITE, buf);
 	err = sysctl_handle_string(oidp, buf, sizeof (buf), req);
-	if (err)
+	if (err || req->newptr == NULL)
 		return (err);
 	return (spa_taskq_param_set(ZIO_TYPE_WRITE, buf));
 }

--- a/module/zfs/spa.c
+++ b/module/zfs/spa.c
@@ -1345,7 +1345,7 @@ spa_taskq_param_set(zio_type_t t, char *cfg)
 }
 
 static int
-spa_taskq_param_get(zio_type_t t, char *buf)
+spa_taskq_param_get(zio_type_t t, char *buf, boolean_t add_newline)
 {
 	int pos = 0;
 
@@ -1363,7 +1363,8 @@ spa_taskq_param_get(zio_type_t t, char *buf)
 		sep = " ";
 	}
 
-	buf[pos++] = '\n';
+	if (add_newline)
+		buf[pos++] = '\n';
 	buf[pos] = '\0';
 
 	return (pos);
@@ -1381,7 +1382,7 @@ spa_taskq_read_param_set(const char *val, zfs_kernel_param_t *kp)
 static int
 spa_taskq_read_param_get(char *buf, zfs_kernel_param_t *kp)
 {
-	return (spa_taskq_param_get(ZIO_TYPE_READ, buf));
+	return (spa_taskq_param_get(ZIO_TYPE_READ, buf, TRUE));
 }
 
 static int
@@ -1395,7 +1396,7 @@ spa_taskq_write_param_set(const char *val, zfs_kernel_param_t *kp)
 static int
 spa_taskq_write_param_get(char *buf, zfs_kernel_param_t *kp)
 {
-	return (spa_taskq_param_get(ZIO_TYPE_WRITE, buf));
+	return (spa_taskq_param_get(ZIO_TYPE_WRITE, buf, TRUE));
 }
 #else
 /*
@@ -1410,7 +1411,7 @@ spa_taskq_read_param(ZFS_MODULE_PARAM_ARGS)
 	char buf[SPA_TASKQ_PARAM_MAX];
 	int err;
 
-	(void) spa_taskq_param_get(ZIO_TYPE_READ, buf);
+	(void) spa_taskq_param_get(ZIO_TYPE_READ, buf, FALSE);
 	err = sysctl_handle_string(oidp, buf, sizeof (buf), req);
 	if (err || req->newptr == NULL)
 		return (err);
@@ -1423,7 +1424,7 @@ spa_taskq_write_param(ZFS_MODULE_PARAM_ARGS)
 	char buf[SPA_TASKQ_PARAM_MAX];
 	int err;
 
-	(void) spa_taskq_param_get(ZIO_TYPE_WRITE, buf);
+	(void) spa_taskq_param_get(ZIO_TYPE_WRITE, buf, FALSE);
 	err = sysctl_handle_string(oidp, buf, sizeof (buf), req);
 	if (err || req->newptr == NULL)
 		return (err);

